crates/holochain/src/core/ribosome/host_fn/get_agent_activity.rs

```diff
@@ -53,6 +59,54 @@ pub fn get_agent_activity(
                         wasm_error!(WasmErrorInner::Host(cascade_error.to_string()))
                     })?;
 
+                #[cfg(feature = "unstable-warrants")]
+                if !activity.warrants.is_empty() {
```

Comment:
So it looks like this is the only host function that both reads and writes data at this level. All others do some secondary writing (i.e. caching data that was fetched) at the cascade level.

I know that state handling is a bit of a mess between this crate, holochain_cascade, and holochain_state. My suggestion would be to keep it consistent for now (until we get to that refactoring), and move this into the `CascadeImpl`.

Reply:
Edit: I see that we do directly insert warrants into the cache in the must_get_agent_activity and must_get_valid_record host fns. That doesn't seem right to me either, but I'll defer to you if this is better saved for the state refactoring work.

Reply:
I liked your suggestion, but the cascade only has a read handle on the DHT database. Also it can't trigger sys validation, so the host fn would have to somehow know that a warrant has successfully been inserted into the DHT to trigger the sys validation.

Reply:
> cascade only has a read handle on the DHT database

I'm not sure where to find this in the codebase, but is that wrong now? Since you need to write warrants directly to cache db I'd think that should be changed to have write access to cache? Should we make that change or defer this?

The cascade fn could return how many warrants it inserted and then you could trigger sys validation here based on that.

If you're just thinking this will all be refactored away so not worth addressing here, I'd defer to you. Don't want to hold up the PR if you want to just resolve this convo.